token/program-2022/src/extension/metadata_pointer/instruction.rs

```diff
@@ -0,0 +1,122 @@
+use {
+    crate::{
+        check_program_account,
+        instruction::{encode_instruction, TokenInstruction},
+        pod::OptionalNonZeroPubkey,
+    },
+    bytemuck::{Pod, Zeroable},
+    num_enum::{IntoPrimitive, TryFromPrimitive},
+    solana_program::{
+        instruction::{AccountMeta, Instruction},
+        program_error::ProgramError,
+        pubkey::Pubkey,
+    },
+    std::convert::TryInto,
+};
+
+/// Metadata pointer extension instructions
+#[derive(Clone, Copy, Debug, PartialEq, IntoPrimitive, TryFromPrimitive)]
+#[repr(u8)]
+pub enum MetadataPointerInstruction {
+    /// Initialize a new mint with a metadata pointer
+    ///
+    /// Fails if the mint has already been initialized, so must be called before
+    /// `InitializeMint`.
+    ///
+    /// The mint must have exactly enough space allocated for the base mint (82
+    /// bytes), plus 83 bytes of padding, 1 byte reserved for the account type,
+    /// then space required for this extension, plus any others.
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    ///   0. `[writable]` The mint to initialize.
+    ///
+    /// Data expected by this instruction:
+    ///   `crate::extension::metadata_pointer::instruction::InitializeInstructionData`
+    ///
+    Initialize,
+    /// Update the metadata pointer address. Only supported for mints that
+    /// include the `MetadataPointer` extension.
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    ///   * Single authority
+    ///   0. `[writable]` The mint.
+    ///   1. `[signer]` The metadata pointer authority.
+    ///
+    ///   * Multisignature authority
+    ///   0. `[writable]` The mint.
+    ///   1. `[]` The mint's metadata pointer authority.
+    ///   2. ..2+M `[signer]` M signer accounts.
+    ///
+    /// Data expected by this instruction:
+    ///   `crate::extension::metadata_pointer::instruction::UpdateInstructionData`
+    ///
+    Update,
+}
```

Comment:
I think we _may_ end-up accidentally kneecapping collections via this interface, but otherwise looks good!

We can always ship a new interface for Metadata forks with collection mgmt

cc @austbot

Reply:
Sorry, I don't know too much about collections -- how would this affect them? Are you thinking of a situation in which some mints in a collection get their metadata changed so they're no longer part of a collection?

Reply:
Personally I think it's worth fleshing out your concern @ngundotra - I don't like the sound of two metadata interfaces right out of the gate, only one of which is forkable

Reply:
Metaplex's Token Metadata has the ability to set & verify collections of NFTs (grouping ID + some collection level metadata) but this current Metadata interface doesn't have any setup for that.

I think it's fine to push this out + then see what people naturally do to create collections/grouping on top.

We might get positively surprised シ

Reply:
I dug through the Metaplex code, and here's how it currently works:

* there's a collection parent mint, with its own metadata
* child mints that are part of this collection have a pointer in their metadata to the parent *mint* address
* child mints are verified by checking that the parent mint's metadata update authority signs a transaction

Depending on what we want, there's a few ways to go about it.

We can add a "collection" interface, which does the following:

* init / update collection parent -- sets in an authority to an account
* add / remove child -- requires collection parent to sign, writes or unsets the collection parent address in the child

You just need two structs: `CollectionParent` and `CollectionChild`.

```
struct CollectionParent {
    authority: Pubkey,
    /// any other fields that you might want, maybe all of the children pubkeys
}

struct CollectionChild {
    parent: Pubkey
}
```

Note that there's *nothing* about tokens or metadata in here. A program that implements just the token metadata interface can tack on these TLV structures for parent and child, and it'll work very similarly to metaplex. Rather than writing the parent mint address though, it'll write the parent metadata address.

BUT, and here's where this gets cool: token-2022 can implement this interface too. And in that case, everything will point to mints.

What do you think?

Reply:
So you'd just add the collection pointer into the metadata (or Token2022's mint) the same way you're kind of doing the metadata pointer rn?

> child mints are verified by checking that the parent mint's metadata update authority signs a transaction

And the metadata interface still has an update authority, so this should still be possible for those implementing the collection interface, so that's good.

I think the tricky part here is that MPL Token Metadata owns all the PDA accounts housing metadata/collection data, so there's more flexibility across data.
More specifically, MPL-TM can verify collection members like so:
- You create metadata and point at the collection, saying you are a part of the collection (unverified)
- You send an instruction to MPL-TM (`VerifyCollection`), and it validates the update authority's signature and **changes the data field in the metadata to verified** - see [Metadata](https://docs.rs/mpl-token-metadata/latest/mpl_token_metadata/state/struct.Metadata.html) and [Collection](https://docs.rs/mpl-token-metadata/latest/mpl_token_metadata/state/struct.Collection.html) (from collection field in Metadata)

So you can kind of think of MPL-TM as a program that implements _both_ the `token-metadata` and `token-collection`(or whatever) interfaces.
With that in mind, it's completely possible to replicate MPL-TM collection-metadata mapping **if your program implements both interfaces**. 
However, if you are using one program's collection standard and another program's metadata standard, then what?

Reply:
Ah sorry I said this part wrong:

> child mints are verified by checking that the parent mint's metadata update authority signs a transaction

Should be the parent mint's "collection authority", and that's it, since there's no overlap between metadata and collections with this interface. The two concepts are *totally* distinct.

>  More specifically, MPL-TM can verify collection members like so:

We don't need to worry about verified and unverified, since you must sign with the parent authority in order to write the collection pubkey in the child. Every mint is immediately verified.

Unless there's a workflow where someone wants to add a mint to a collection as unverified. Or where unverified token mints are important. We can add a `verified` bool in the Child struct if needed though, no problem!

>  However, if you are using one program's collection standard and another program's metadata standard, then what?

Then you're a metadata program like metaplex's token-metadata, right? Maybe I'm missing something more profound to this question, sorry.

The whole point is that you can have just a metadata program, you can have just a collection program, and you can have both -- it's all interfaces!